packages/origin-247-transfer/src/transfer.service.ts

```diff
@@ -0,0 +1,211 @@
+import { QueryBus, CommandBus, EventBus } from '@nestjs/cqrs';
+import {
+    GetTransferSitesQuery,
+    IGetTransferSitesQueryResponse
+} from './queries/GetTransferSites.query';
+import { Logger } from '@nestjs/common';
+import {
+    ValidateTransferCommandCtor,
+    VALIDATE_TRANSFER_COMMANDS_TOKEN
+} from './commands/ValidateTransferCommand';
+import { EnergyTransferRequest, TransferValidationStatus } from './EnergyTransferRequest';
+import { ValidatedTransferRequestEvent } from './events';
+
+interface IIssueCommand extends GenerationReadingStoredPayload<unknown> {}
+
+export interface IUpdateValidationStatusCommand {
+    requestId: number;
+    status: TransferValidationStatus;
+    validatorName: string;
+}
+
+@Injectable()
+export class TransferService {
+    private readonly logger = new Logger(TransferService.name);
+
+    constructor(
+        @Inject(CERTIFICATE_SERVICE_TOKEN)
+        private certificateService: CertificateService<unknown>,
+        @Inject(ENERGY_TRANSFER_REQUEST_REPOSITORY)
+        private energyTransferRequestRepository: EnergyTransferRequestRepository,
+        @Inject(VALIDATE_TRANSFER_COMMANDS_TOKEN)
+        private validateCommands: ValidateTransferCommandCtor[],
+        private queryBus: QueryBus,
+        private commandBus: CommandBus,
+        private eventBus: EventBus
+    ) {}
+
+    public async issue(command: IIssueCommand): Promise<void> {
+        const { generatorId, energyValue, transferDate, fromTime, toTime, metadata } = command;
+
+        const sites: IGetTransferSitesQueryResponse | null = await this.queryBus.execute(
+            new GetTransferSitesQuery({ generatorId })
+        );
+
+        if (!sites) {
+            this.logger.log(`No sites queries for ${generatorId}`);
+            return;
+        }
+
+        const request = await this.energyTransferRequestRepository.createNew({
+            buyerAddress: sites.buyerAddress,
+            sellerAddress: sites.sellerAddress,
+            volume: energyValue,
+            generatorId,
+            transferDate
+        });
+
+        const certificate = await this.certificateService.issue({
+            deviceId: generatorId,
+            energyValue: energyValue,
+            fromTime,
+            toTime,
+            userId: sites.sellerAddress,
+            toAddress: sites.sellerAddress,
+            metadata
+        });
+
+        request.updateCertificateId(certificate.id);
+
+        await this.energyTransferRequestRepository.save(request);
+
+        // There is a risk of race condition between `issue` finish and `CertificatePersistedEvent`
+        // If certificate is already persisted (it can be found by service)
+        // Then we should proceed as we received the event
+
+        const isCertificatePersisted = Boolean(
+            await this.certificateService.getById(certificate.id)
+        );
+
+        if (isCertificatePersisted) {
```

Comment:
if(!isCertificatePersisted)

Reply:
It is correct. 

Expected result is that we have issue then persisted. Then issue has the id, and we are waiting for the persisted event.

However if we received persited event first (`persistRequestCertificate`) and `issue` didn't finish, then `persistRequestCertificate` aborts.

So in this case if we finish issueing and certificate is already persisted, we need to push this `persistRequestCertificate` manually (because event was already called).